content/statefulset/statefulset.md

```diff
@@ -0,0 +1,69 @@
+---
+title: "Create StatefulSet"
+date: 2018-08-07T08:30:11-07:00
+weight: 20
+---
+#### Introduction
+StatefulSet consists of serviceName, replicas, template and volumeClaimTemplates.
+**serviceName** is "mysql", headless service we created in previous section, **replicas** is 3, the desired number of pod, **template** is the configuration of pod, **volumeClaimTemplates** is to claim volume for pod based on storageClassName, gp2 that we created in "Define Storageclass" section. [Percona Xtrabackup](https://www.percona.com/software/mysql-database/percona-xtrabackup) is in template to clone source MySQL server to its slaves.
+#### Create StatefulSet
+Copy/Paste the following commands into your Cloud9 Terminal.
+```
+cd ~/environment/templates
+wget https://eksworkshop.com/statefulset/statefulset.files/mysql-statefulset.yml
+```
+Check statefulset "mysql" created by following command.
+```
+kubectl create -f ~/environment/templates/mysql-statefulset.yml
+statefulset "mysql" created
+```
+#### Watch StatefulSet
+Watch the status of statefulset.
+```
+kubectl get -w statefulset
+NAME      DESIRED   CURRENT   AGE
+mysql     3         1         8s
+mysql     3         2         59s
+mysql     3         3         2m
+mysql     3         3         3m
+```
+Open another Cloud9 Terminal and watch the progress of pods creation using the following command. You can see ordered, graceful deployment with a stable, unique name for each pod.
+```
+kubectl get pods -l app=mysql --watch
+NAME      READY     STATUS     RESTARTS   AGE
+mysql-0   0/2       Init:0/2   0          30s
+mysql-0   0/2       Init:1/2   0         35s
+mysql-0   0/2       PodInitializing   0         47s
+mysql-0   1/2       Running   0         48s
+mysql-0   2/2       Running   0         59s
+mysql-1   0/2       Pending   0         0s
+mysql-1   0/2       Pending   0         0s
+mysql-1   0/2       Pending   0         0s
+mysql-1   0/2       Init:0/2   0         0s
+mysql-1   0/2       Init:1/2   0         35s
+mysql-1   0/2       Init:1/2   0         45s
+mysql-1   0/2       PodInitializing   0         54s
+mysql-1   1/2       Running   0         55s
+mysql-1   2/2       Running   0         1m
+mysql-2   0/2       Pending   0         <invalid>
+mysql-2   0/2       Pending   0         <invalid>
+mysql-2   0/2       Pending   0         0s
+mysql-2   0/2       Init:0/2   0         0s
+mysql-2   0/2       Init:1/2   0         32s
+mysql-2   0/2       Init:1/2   0         43s
+mysql-2   0/2       PodInitializing   0         50s
+mysql-2   1/2       Running   0         52s
+mysql-2   2/2       Running   0         56s
+```
```

Comment:
I get errors here:
```
brent:~/environment $ kubectl get pods -l app=mysql --watch
NAME      READY     STATUS     RESTARTS   AGE
mysql-0   2/2       Running    0          1m
mysql-1   0/2       Init:1/2   0          36s
mysql-1   0/2       Init:1/2   0         41s
mysql-1   0/2       Init:Error   0         1m
mysql-1   0/2       Init:1/2   1         1m
mysql-1   0/2       Init:Error   1         1m
mysql-1   0/2       Init:CrashLoopBackOff   1         1m
mysql-1   0/2       Init:1/2   2         1m
mysql-1   0/2       Init:Error   2         1m
mysql-1   0/2       Init:CrashLoopBackOff   2         2m
```

Reply:
This might be the storageclass, gp2 was pre-created at your environment with different parameters. It will work now because I changed storageclassname=mysql-gp2.